src/python/pants/backend/project_info/dependencies.py

```diff
@@ -31,17 +43,74 @@ class DependenciesSubsystem(LineOriented, GoalSubsystem):
         default=False,
         help="Include the input targets in the output, along with the dependencies.",
     )
+    format = EnumOption(
+        default=DependenciesOutputFormat.merged,
+        help="Output format for listing dependencies.",
+    )
 
 
 class Dependencies(Goal):
     subsystem_cls = DependenciesSubsystem
     environment_behavior = Goal.EnvironmentBehavior.LOCAL_ONLY
 
 
-@goal_rule
-async def dependencies(
-    console: Console, addresses: Addresses, dependencies_subsystem: DependenciesSubsystem
-) -> Dependencies:
+async def list_dependencies_as_json(
+    addresses: Addresses, dependencies_subsystem: DependenciesSubsystem, console: Console
+) -> None:
+    """Get dependencies for given addresses and list them in the console in JSON.
+
+    Note that `--closed` option is ignored as it doesn't make sense to duplicate source address in
+    the list of its dependencies.
+    """
+    # NB: We must preserve target generators for the roots, i.e. not replace with their
+    # generated targets.
+    target_roots = await Get(UnexpandedTargets, Addresses, addresses)
+    # NB: When determining dependencies, though, we replace target generators with their
+    # generated targets.
+    if dependencies_subsystem.transitive:
+        transitive_targets_group = await MultiGet(
+            Get(
+                TransitiveTargets,
+                TransitiveTargetsRequest(
+                    (address,), should_traverse_deps_predicate=AlwaysTraverseDeps()
+                ),
+            )
+            for address in addresses
+        )
+
+        iterated_targets = []
+        for transitive_targets in transitive_targets_group:
+            iterated_targets.append(
+                sorted([str(tgt.address) for tgt in transitive_targets.dependencies])
+            )
+
+    else:
+        dependencies_per_target_root = await MultiGet(
+            Get(
+                Targets,
+                DependenciesRequest(
+                    tgt.get(DependenciesField),
+                    should_traverse_deps_predicate=AlwaysTraverseDeps(),
+                ),
+            )
+            for tgt in target_roots
+        )
+
+        iterated_targets = []
+        for targets in dependencies_per_target_root:
+            iterated_targets.append(sorted([str(tgt.address) for tgt in targets]))
+
+    # the assumption is that when iterating the targets and sending dependency requests
+    # for them, the lists of dependencies are returned in the very same order
```

Comment:
Say we have 2 addresses, `A1` and `A2`, and two `Get` requests are sent while iterating the addresses. Now we got a tuple of two lists from the `MultiGet`, `[D1, D2]` and `[D3, D4]`. We assume that since the requests were passed in a particular order, this is the order in which the results are returned as well, so we map `A1` to `[D1, D2]` and `A2` to `[D3, D4]`.

If this is not the case, we'd need to create a mapping manually and fill it in a `for` loop, iterating the addresses individually, but this will have a performance hit.